### PR TITLE
Add unit test cases for System.Runtime.Loader.AssemblyLoadContext

### PR DESCRIPTION
**Pull request by Sridhar-MS**
Usage of TrustedPlatformAssemblies and AssemblyLoadContext is mutually exclusive, you cannot use both. Since the premise is that you either want to use the default binding mechanism (via coreclr TPA binder) or supply your own (via AssemblyLoadContext) for your own assemblies. On the other hand corerun blindly adds all the assemblies (including test assemblies) in the current directory to the TPA list (and hence cannot be loaded via custom load contexts). To workaround this issue the test assembly which is loaded by a custom load context is embedded as a resource to the unit test assembly. This resource is extracted at runtime and prevents corerun from adding it to the TPA list.

### CODE REVIEW

File: src/System.Runtime.Loader/tests/AssemblyLoadContextTest.cs [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using System.Threading.Tasks;
+
+namespace System.Runtime.Loader.Tests
+{
+    public class AssemblyLoadContextTest
+    {
+        const string TESTASSEMBLY = "System.Runtime.Loader.TestAssembly";
```
Comment (stephentoub) [inline src/System.Runtime.Loader/tests/AssemblyLoadContextTest.cs]: The guideline for the repo would have this be cased as ```TestAssembly```.

File: src/System.Runtime.Loader/tests/AssemblyLoadContextTest.cs [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+        public static void GetLoadContextTest_ValidUserAssembly()
+        {
+            var asmName = new AssemblyName(TESTASSEMBLY);
+            var loadContext = new ResourceAssemblyLoadContext();
+            loadContext.LoadBy = LoadBy.Stream;
+
+            var asm = loadContext.LoadFromAssemblyName(asmName);
+            var context = AssemblyLoadContext.GetLoadContext(asm);
+
+            Assert.True(context == loadContext);
```
Comment (stephentoub) [inline src/System.Runtime.Loader/tests/AssemblyLoadContextTest.cs]: Assert.Equal?

**Comment by stephentoub** [comment]
Several tests failed in CI on Linux:
```
System.Runtime.Loader.Tests.AssemblyLoadContextTest.LoadAssemblyByPath_ValidUserAssembly [FAIL]
20:05:21       System.IO.FileLoadException : Could not load file or assembly 'System.Runtime.Loader.TestAssembly, Version=999.999.999.999, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies.
20:05:21       Stack Trace:
20:05:21             at System.Runtime.Loader.AssemblyLoadContext.LoadFromPath(IntPtr ptrNativeAssemblyLoadContext, String ilPath, String niPath, ObjectHandleOnStack retAssembly)
20:05:21             at System.Runtime.Loader.AssemblyLoadContext.LoadFromAssemblyPath(String assemblyPath)
20:05:21             at System.Runtime.Loader.Tests.ResourceAssemblyLoadContext.Load(AssemblyName assemblyName)
20:05:21             at System.Runtime.Loader.AssemblyLoadContext.LoadFromAssemblyName(AssemblyName assemblyName)
20:05:21             at System.Runtime.Loader.Tests.AssemblyLoadContextTest.LoadAssemblyByPath_ValidUserAssembly()
20:05:21    System.Runtime.Loader.Tests.AssemblyLoadContextTest.LoadAssemblyByStream_ValidUserAssembly [FAIL]
20:05:21       System.IO.FileLoadException : Could not load file or assembly 'System.Runtime.Loader.TestAssembly, Version=999.999.999.999, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies.
20:05:21       Stack Trace:
20:05:21             at System.Runtime.Loader.AssemblyLoadContext.LoadFromStream(IntPtr ptrNativeAssemblyLoadContext, IntPtr ptrAssemblyArray, Int32 iAssemblyArrayLen, IntPtr ptrSymbols, Int32 iSymbolArrayLen, ObjectHandleOnStack retAssembly)
20:05:21             at System.Runtime.Loader.AssemblyLoadContext.LoadFromStream(Stream assembly, Stream assemblySymbols)
20:05:21             at System.Runtime.Loader.Tests.ResourceAssemblyLoadContext.Load(AssemblyName assemblyName)
20:05:21             at System.Runtime.Loader.AssemblyLoadContext.LoadFromAssemblyName(AssemblyName assemblyName)
20:05:21             at System.Runtime.Loader.Tests.AssemblyLoadContextTest.LoadAssemblyByStream_ValidUserAssembly()
20:05:21    System.Runtime.Loader.Tests.AssemblyLoadContextTest.GetLoadContextTest_ValidUserAssembly [FAIL]
20:05:21       System.IO.FileLoadException : Could not load file or assembly 'System.Runtime.Loader.TestAssembly, Version=999.999.999.999, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies.
20:05:21       Stack Trace:
20:05:21             at System.Runtime.Loader.AssemblyLoadContext.LoadFromStream(IntPtr ptrNativeAssemblyLoadContext, IntPtr ptrAssemblyArray, Int32 iAssemblyArrayLen, IntPtr ptrSymbols, Int32 iSymbolArrayLen, ObjectHandleOnStack retAssembly)
20:05:21             at System.Runtime.Loader.AssemblyLoadContext.LoadFromStream(Stream assembly, Stream assemblySymbols)
20:05:21             at System.Runtime.Loader.Tests.ResourceAssemblyLoadContext.Load(AssemblyName assemblyName)
20:05:21             at System.Runtime.Loader.AssemblyLoadContext.LoadFromAssemblyName(AssemblyName assemblyName)
20:05:21             at System.Runtime.Loader.Tests.AssemblyLoadContextTest.GetLoadContextTest_ValidUserAssembly()
```

File: src/System.Runtime.Loader/tests/System.Runtime.Loader.TestAssembly/project.lock.json [hunk 3]
```diff
@@ -0,0 +1,381 @@
+{
+  "locked": false,
```
Comment (stephentoub) [inline src/System.Runtime.Loader/tests/System.Runtime.Loader.TestAssembly/project.lock.json]: Should be changed to true

**Comment by stephentoub** [comment]
Thanks, Sridhar.  A few issues (in particular the failing tests), but otherwise looks good.  Good to see some tests starting to come online for this stuff.

**Comment by Sridhar-MS** [comment]
Thanks Stephen! I will look into the linux failure. It did pass on my linux VM.

File: src/System.Runtime.Loader/tests/AssemblyLoadContextTest.cs (hunk 1 above)
Comment (mellinoe) [inline src/System.Runtime.Loader/tests/AssemblyLoadContextTest.cs]: Also, visibility should be explicitly specified.

**Comment by mmitche** [comment]
@dotnet-bot test this please

**Comment by Sridhar-MS** [comment]
System.Runtime.Loader.TestAssembly.dll gets copied into the test folder "System.Runtime.Loader.Tests/dnxcore50" even though it is not referenced by System.Runtime.Loader.Tests.dll. Therefore it gets added to the TPA list by corerun and fails the test.

This happens only on linux.

**Comment by mmitche** [comment]
Where can the TestAssembly come from?  I'm thinking it made it into the test overlay but I'm not sure how.

**Comment by Sridhar-MS** [comment]
I am looking at the linux build logs at - http://dotnet-ci.cloudapp.net/job/dotnet_corefx_linux_debug_prtest/2676/console

```
System.Runtime.Loader.TestAssembly -> d:\j\workspace\dotnet_corefx_linux_debug_prtest\bin\Linux.AnyCPU.Debug\System.Runtime.Loader.TestAssembly\System.Runtime.Loader.TestAssembly.dll
  System.Runtime.Loader.Tests -> d:\j\workspace\dotnet_corefx_linux_debug_prtest\bin\Linux.AnyCPU.Debug\System.Runtime.Loader.Tests\System.Runtime.Loader.Tests.dll
```

Both the assemblies are getting copied to their own directories.

**Comment by Sridhar-MS** [comment]
@mmitche I made the fix in run-tests.sh to not copy test libs into the 'overlay' directory. Also previously a bunch of test files where getting copied
```
System.Diagnostics.FileVersionInfo.TestAssembly.dll
System.Runtime.Loader.TestAssembly.dll
XDocument.Test.ModuleCore.dll
XmlCoreTest.dll
```

We must not be seeing them again in the overlay directory.

**Comment by mmitche** [comment]
@Sridhar-MS LGTM

**Comment by stephentoub** [comment]
LGTM

**Pull request merged**